client/src/components/SigninContainer.js

```diff
@@ -12,89 +12,64 @@ import {
-      case "textChange":
-        return {
-          ...state,
-          [action.name]: action.value,
-          emailError: false,
-          emailValidationError: false,
-          pwdError: false,
-          pwdValidationError: false,
-        };
-      case "error":
-        return { ...state, [action.error]: true };
-      case "update":
-        return state;
-      default:
-        throw new Error();
-    }
-  };
+	const [email, setEmail] = useState("");
+	const [pwd, setPwd] = useState("");
+	const [emailClientError, setEmailClientError] = useState(false);
+	const [emailServerError, setEmailServerError] = useState(false);
+	const [pwdClientError, setPwdClientError] = useState(false);
+	const [pwdServerError, setPwdServerError] = useState(false);
 
-  const [state, dispatch] = useReducer(reducer, initialState);
-
-  useEffect(() => {
-    dispatch({ type: "update" });
-  }, [state]);
-
-  const handleInputChange = (event) => {
-    dispatch({
-      type: "textChange",
-      name: event.target.name,
-      value: event.target.value,
-    });
-  };
+	const handleEmail = (event) => {
+		setEmail(event.target.value);
+		setEmailClientError(false);
+		setEmailServerError(false);
+	};
+	const handlePwd = (event) => {
+		setPwd(event.target.value);
+		setPwdClientError(false);
+		setPwdServerError(false);
+	};
 
-  const checkEmail = () => {
-    const emailPattern = new RegExp(`[a-z0-9._%+-]+@[a-z0-9.-]+\\.[a-z]{2,}$`);
-    const emailTest = emailPattern.test(state.emailSignin);
-    return emailTest;
-  };
+	const checkEmail = () => {
+		const emailPattern = new RegExp(`[a-z0-9._%+-]+@[a-z0-9.-]+\\.[a-z]{2,}$`);
+		return emailPattern.test(email);
+	};
 
 	const handleSubmit = async (event) => {
 		event.preventDefault();
-		const { emailSignin, pwdSignin } = state;
 		const userInfo = {
-			email: emailSignin.trim().toLowerCase(),
-			password: pwdSignin.trim(),
+			email: email.trim().toLowerCase(),
+			password: pwd.trim(),
 		};
-		if (checkEmail() && pwdSignin) {
-			const loginResponse = await (await fetch("/api/users/login", {
+		if (checkEmail() && pwd) {
+			const loginResponse = await fetch("/api/users/login", {
 				method: "POST",
 				headers: {
 					"Content-Type": "application/json",
 				},
 				body: JSON.stringify(userInfo),
-			})).json();
-			if (loginResponse.status === "success") {
-				dash(); // redirect to user's dashboard
-			} else if ("email" in loginResponse) {
-				dispatch({ type: "error", error: "emailError" });
-			} else if ("password" in loginResponse) {
-				dispatch({ type: "error", error: "pwdError" });
-			} else if ("errors" in loginResponse) {
-				if ("email" in loginResponse.errors) {
-					dispatch({ type: "error", error: "emailValidationError" });
-				}
+			});
+			const loginData = await loginResponse.json();
+			if (loginData.status === "success") {
+				// need to close modal
+				dash(); // redirect to explore page
+				dispatch({ type: "AUTHENTICATED", payload: loginData.data });
```

Comment:
If the redirect happens, won't the dispatch be skipped?

Reply:
No, the dispatch is not skipped. I checked context. However, I can change the order.